### ISSUE TITLE
Property initialization false positive (MixedPropertyTypeCoercion)

### ISSUE THREAD

**Issue by enumag** [vimeo/psalm#2755]
Just trying out psalm and one of the very common things that psalm wrongly reports is this:

```
ERROR: MixedPropertyTypeCoercion - src/ReadModel/Main/Projection/Security/UserProjection.php:77:18 - $this->roles expects 'Doctrine\Common\Collections\Collection<int, App\ReadModel\Main\Projection\Security\UserRoleProjection>',  parent type `Doctrine\Common\Collections\ArrayCollection<mixed, empty>` provided
                $this->roles = new ArrayCollection();

```

Code:

```php
/**
 * @ORM\Entity()
 */
class UserProjection
{
	// ...

	/**
	 * @var Collection<int, UserRoleProjection>
	 * @ORM\OneToMany(...)
	 */
	private $roles;

	public function __construct(Identity $id)
	{
		$this->roles = new ArrayCollection();
	}
}
```

**Comment by muglug** [comment]
Yup, this is a weakness of just having generics in docblocks – there are some things that cannot be inferred.

Instead you need to use `/** @var Collection<UserRoleProjection> */` docblock above the assignment.

If/when PHP gets generics in the language, you'll be able to write `new Collection<UserRoleProjection>()` and Psalm will be content.

**Comment by enumag** [comment]
> Yup, this is a weakness of just having generics in docblocks – there are some things that cannot be inferred.

PHPStan handles this correctly so this is simply a weakness of Psalm.

> Instead you need to use `/** @var Collection<UserRoleProjection> */` docblock above the assignment.

lol... no way I'm duplicating the type info in such way - psalm is supposed to help me detect issues, not make me write useless code

> If/when PHP gets generics in the language, you'll be able to write `new Collection<UserRoleProjection>()` and Psalm will be content.

Unfortunately that won't happen for at least several years so "wait for PHP support" is not an argument you should be using.

**Comment by muglug** [comment]
> PHPStan handles this correctly so this is simply a weakness of Psalm.

With respect, PHPStan ignores a whole host of issues that Psalm pays attention to. Generally all of Psalm's issues that have a `Mixed` prefix are ignored by PHPStan, and if you don't care about those you can ignore them via the config (`totallyTyped="false"` is a shortcut).

> Unfortunately that won't happen for at least several years so "wait for PHP support" is not an argument you should be using.

It's not an argument – it's just reality.

---

I'd also encourage you to be a little less aggressive – this isn't a fight or an argument, I'm just a guy at a keyboard explaining how I've decided to design a generic templating system in a language that doesn't support it syntactically.

**Comment by muglug** [comment]
Here's an example of when that issue could lead to a bug: https://phpstan.org/r/b1d0bdb1-61a5-41e5-af09-3ba63de7cd26

Here's how Psalm uses that `MixedPropertyTypeCoercion` warning to tell you that something is possibly bad: https://psalm.dev/r/ed39162b1d

**Comment by enumag** [comment]
> With respect, PHPStan ignores a whole host of issues that Psalm pays attention to. Generally all of Psalm's issues that have a Mixed prefix are ignored by PHPStan, and if you don't care about those you can ignore them via the config (totallyTyped="false" is a shortcut).

So far most of the reports of Mixed* Psalm reported were either things I can't really do anything about or worse just straight out false positives. So there were the first thing for me to suppress.

> I'd also encourage you to be a little less aggressive – this isn't a fight or an argument, I'm just a guy at a keyboard explaining how I've decided to design a generic templating system in a language that doesn't support it syntactically.

Not sure why you thought of my comment as aggressive, it certainly wasn't meant that way. My apology if it sounds aggressive. I'm just trying to understand the issue.

> Here's an example of when that issue could lead to a bug: https://phpstan.org/r/b1d0bdb1-61a5-41e5-af09-3ba63de7cd26

That's a completely different case. In this code the collection already starts with some data given to the constructor so it already has a type before being assigned to the property. In the case I reported the collection holds no data yet and is therefore valid.

Perhaps such case is difficult to understand for Psalm but it doesn't change the fact that Psalm is sometimes too aggressive about reporting things it doesn't understand correctly. Overall my experience with psalm so far is that it reports like 10x times more false positives than actual issues, causing more pain than help.

**Comment by muglug** [comment]
> Overall my experience with psalm so far is that it reports like 10x times more false positives

Please ticket any false-positives you find, especially if you come across something that PHPStan is able to infer is indeed correct (vs something that PHPStan doesn’t care about as above).

**Comment by enumag** [comment]
> especially if you come across something that PHPStan is able to infer is indeed correct (vs something that PHPStan doesn’t care about as above).

It's impossible for me to tell these two categories apart.

**Comment by muglug** [comment]
> It's impossible for me to tell these two categories apart.

For example:

```php
final class C {
  private string $s;

  public static function make(string $s) : self {
    $c = new self();
    $c->s = $s;
    
    return $c;
  }
}
```

Psalm complains that `C::$s` is missing a constructor, because Psalm performs a lot of property type checks under the assumption that `__construct` should be used for property initialisation.

You might consider the above a false-positive because the code works. PHPStan doesn't care about property initialisation, so it won't give you an error.

**Comment by muglug** [comment]
> In the case I reported the collection holds no data yet and is therefore valid.

[Here's another example](https://phpstan.org/r/a2dc9dd2-7526-4734-b418-0830bc2930b7) of where PHPStan wouldn't report a bug in your code. This isn't a bug in PHPStan – it's consistent in its treatment. But it does show the limits of its ability to find bugs, at least currently.
